src/dmd/expression.d

```diff
@@ -4828,7 +4828,29 @@ extern (C++) final class DotVarExp : UnaExp
                                 if (!onlyUnion)
                                     v.ctorinit = false;
                                 if (modifyLevel == Modifiable.initialization)
+                                {
+                                    // https://issues.dlang.org/show_bug.cgi?id=22118
+                                    // v is a union type field that was assigned
+                                    // a variable, therefore it counts as initialization
+                                    if (v.ctorinit)
+                                        return Modifiable.initialization;
+
+                                    // if v is of struct type, then we need to undo
+                                    // the information that was stored in fieldInit
+                                    // to get rid of the wrong error messages
+                                    auto ad = sc.func.isMemberDecl();
+                                    assert(ad);
+                                    size_t i;
+                                    for (i = 0; i < sc.ctorflow.fieldinit.length; i++)
+                                    {
+                                        if (ad.fields[i] == v)
+                                            break;
+                                    }
```

Comment:
@RazvanN7 This code has a bug - what happens if v is not found? Then `i` is out of bounds for line 4849. Better would be:
```
foreach (i, ref fieldInit; sc.ctorflow.fieldinit[])
{
    if (ad.fields[i] == v)
    {
        fieldInit.csx &= CSC,none;
        ... blah blah ...
       return Modifiable.yes;
    }
}
assert(0);
```